pkg/controllers/managementuser/rbac/crtb_handler_test.go

```diff
@@ -143,17 +181,203 @@ func TestSyncCRTB(t *testing.T) {
 			} else {
 				require.NoError(t, err)
 			}
+
+			require.Len(t, test.crtb.Status.Conditions, 1)
+			require.Equal(t, test.condition, rcOf(test.crtb.Status.Conditions[0]))
+		})
+	}
+}
+
+func TestReconcileCRTBUserClusterLabels(t *testing.T) {
```

Comment:
I don't see a test here for removing "resolved" conditions?

Reply:
Can you explain more what you mean by `resolved` condition ? Do you mean `reconciled` ?
Could you provide references into the package code which is not covered ?

Reply:
Sorry, cases where it rereconciles, and the error condition is no longer valid?

Reply:
Also note that IIUC conditions are never removed, only added. I.e. the main status contains the current status of the resource, and the conditions record the history of reconciliation attempts until success at the end.

Reply:
If you look at the status field of `global_roles` you will see that @JonCrowther did the same there, conditions are only added, never removed.

Reply:
To clarify, you're saying that if the state that leads to the condition is resolved and is no longer valid, the condition remains in the status, until _all_ conditions are resolved?

Reply:
IIUC even when all issues are resolved and reconciliation succeeds this only means that a success condition is added at the end of the condition array. Note that conditions have a time associated with them (`LastTransitionTime`).

Reply:
Looking at the system CRTB and GR from a local rancher run I note only successful stati for the varying reconcile functions. The problem is that I cannot say if there had been no failures at all, or if some intermediates failures got implicitly removed.

I.e. I do not know if the controller clears the condition array before it enters the sync/reconcile code.
Because if it does you would automatically get what you asked for while the sync code still only has to add conditions.

... Trying to find docs about the semantics and treatment of conditions was not very successful yet

Reply:
Resolved. Look at https://github.com/rancher/rancher/pull/45269/files#diff-3f8fdf585fe7dae65e27976600c8b9596a5f87ddf9208993f29ad6a8e9c048caR415

The `setCRTBAsInProgress` is the first thing done by the reconcile functions.
This clears the condition array. I.e. ensures that the rest start with an empty array.

The condition you see in the resource are the conditions from the last reconcile attempt, and nothing else.

Reply:
Well, lets see if `crtbclientM` is mockable, and if yes, add tests for the `setCRTBAs...` functions.

Reply:
Unable to locate mocks for `crtbClientM` ... Is a generated thing, and does not seem to have mocks generated for it. ... Checking the GR code I worked from it is not testing the `setGRAs...` functions either, so no help from that area either.

Reply:
@JonCrowther What are your thoughts here ?

Reply:
Maybe it could live in another test function where you run the reconcile function twice? That way you could call it once with mock functions that cause Failed statuses, check that the conditions exist, then update the mocks to Pass and check the Failed status conditions all get removed?

Alternatively, you could add a step in this test function that always runs a second time and verifies that the Failed statuses get removed. Then you'd essentially have a test ensuring every Failed status condition gets removed properly. Might be a little overkill though.

Reply:
The test functions test the internal functions, i.e. `reconcileBindings`, `reconcileSubject`, etc.
These do not remove conditions. The removal is in the internal `setCRTBAsInProgress` which is not tested,
and called before any of the `reconcile...` functions, from the main functions, i.e. `Create`, `Updated`.
Which we do not test.

Regardless of if we try to test `Create`, etc. or `setCRTBAsInProgress`, we will need a mock of the `crtbClientM` field, which is a `("github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3").ClusterRoleTemplateBindingController`. For which I have not found mocks.

Reply:
You don't need to mock much, you could follow the Go guidelines

> Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values. 

So, you could change this...

```go
type clusterRoleTemplateClient interface {    
    UpdateStatus(*v3.ClusterRoleTemplateBinding) (*v3.ClusterRoleTemplateBinding, error)    
}  

type crtbLifecycle struct {    
   ...
    crtbClientM clusterRoleTemplateClient    
}
```

Which would make mocking the interface fairly trivial?

Reply:
Oh. Localized micro mocking through local interface, and alt implementation in the testsuite.
Seems I was very much overthinking this way over much.
Lets try this.

Reply:
Works. Addressed with commit fa2aeb1220a4dae746930f99f746301d4644abc9